### ISSUE TITLE
for-of loops in async functions are incorrectly scoped

### ISSUE THREAD

**Issue by matchu** [swc-project/swc#2624]
### Describe the bug

When you declare the loop variable of a for-of loop with `const`, functions declared during a loop iteration should be bound to that iteration's scope.

However, when the loop appears directly inside an async function, and when targeting ES5, SWC will compile the loop variable to be function-scoped instead of loop-scoped. This can cause incorrect behavior.

### Input code

```typescript
async function runCounters() {
  for (const idebug of [1, 2, 3]) {
    setTimeout(() => console.log(`idebug = ${idebug}`), 100);
  }
}
```

### Playground link

https://play.swc.rs/?version=1.2.106&code=H4sIAAAAAAAAAy3LsQqDMBRG4V3wHf7BIYFQtB3FLr5CtyLYpokE7L2Q5A5F8u6t1O0M53ukD1l4IZsDE6LQyELZxaQ0troCPEcoy5Qywss9ZQF73DuDs8FlOiYguXwLb8eS1U8OV%2ByEV3daeVHzIQc02z%2FLrA26ttX9zktdlf4LmGEAsYsAAAA%3D&config=H4sIAAAAAAAAA0WMTQrEIAxG75K1286id5hDBCctFv9IUhgR714tlu7C915ehUMsrBUyshCPS0pU%2FMMKZAOKZZcVTNf6tKEXagYUeScdiiyd%2BZSEJjUQXHRbGSWbQmYSeRHG3T9m66GQfucYKmjJdAc%2F0N7G%2FHPynaLySe0Cj9ke9LUAAAA%3D

### Expected behavior

When we call `runCounters()`, we should see the following output:
```
idebug = 1
idebug = 2
idebug = 3
```

However, after compiling with SWC to ES5, `runCounters()` instead yields the following:
```
idebug = 3
idebug = 3
idebug = 3
```

Here's the relevant snippet of the output code:
```js
_runCounters = _asyncToGenerator(regeneratorRuntime.mark(function _callee() {
    var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, idebug;
    return regeneratorRuntime.wrap(function _callee$(_ctx) {
        while(1)switch(_ctx.prev = _ctx.next){
            case 0:
                _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
                _ctx.prev = 1;
                for(_iterator = [
                    1,
                    2,
                    3
                ][Symbol.iterator](); !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true){
                    idebug = _step.value;
                    setTimeout(function() {
                        return console.log("idebug = ".concat(idebug));
                    }, 100);
                }
                _ctx.next = 9;
                break;
            case 5:
                // ...omitted additional switch cases...
```

Observe that `idebug` is declared in the `var` group at the top of the `_callee` function, alongside the iteration state. This means that, as iteration proceeds and the iteration state is mutated, the `idebug` variable that our timeout functions reference will be mutated, too.

### Version

1.2.106

### Additional context

I first noticed this when I discovered a noteworthy bug when moving an app to Next.js 12! 😅 Excited for the speed improvements though, big thanks to everyone working on this!! 💖

**Comment by matchu** [comment]
For now, I'm working around this by wrapping my loop inside an immediately-invoked synchronous function!

```js
const runCounters = async () => {
  (() => {
    for (const idebug of [1, 2, 3]) {
      setTimeout(() => console.log(`i = ${idebug}`), 100);
    }
  })();
};
```

(In this simple example, the better fix would be to not declare the function with `async`, because we're not actually using `async` functionality. But in a function where `async` matters, this workaround seems to do the trick!)

**Comment by kdy1** [comment]
Reopening as https://github.com/swc-project/swc/pull/2740 caused a serious bug.

**Comment by joshyudeep** [comment]
@matchu thanks for reporting this -- I'm running into similar issues. 

@kdy1 any updates on this? 

Thanks 🙏 🙏 🙏

**Comment by Austaras** [comment]
Already fixed by #4029

**Comment by kdy1** [comment]
Thanks!
